nano/node/node.cpp

```diff
@@ -1120,72 +1118,119 @@ nano::signature_checker::~signature_checker ()
 {
-	/* Verifications is vector if signatures check results
-	 validate_message_batch returing "true" if there are at least 1 invalid signature */
-	auto code (nano::validate_message_batch (check_a.messages, check_a.message_lengths, check_a.pub_keys, check_a.signatures, check_a.size, check_a.verifications));
+	/* Returns false if there are at least 1 invalid signature */
+	auto code (nano::validate_message_batch (check_a.messages + start_index, check_a.message_lengths + start_index, check_a.pub_keys + start_index, check_a.signatures + start_index, size, check_a.verifications + start_index));
 	(void)code;
-	release_assert (std::all_of (check_a.verifications, check_a.verifications + check_a.size, [](int verification) { return verification == 0 || verification == 1; }));
-	check_a.promise->set_value ();
+
+	return std::all_of (check_a.verifications + start_index, check_a.verifications + start_index + size, [](int verification) { return verification == 0 || verification == 1; });
 }
 
-void nano::signature_checker::run ()
+void nano::signature_checker::verify_threaded (nano::signature_check_set & check_a)
 {
-	nano::thread_role::set (nano::thread_role::name::signature_checking);
-	std::unique_lock<std::mutex> lock (mutex);
-	started = true;
+	constexpr size_t batch_size = 256;
+	size_t overflow = check_a.size % batch_size;
+	size_t full_batches = check_a.size / batch_size;
 
-	lock.unlock ();
-	condition.notify_all ();
-	lock.lock ();
+	auto batches = full_batches + ((overflow != 0) ? 1 : 0);
+	auto task = std::make_shared<Task> (check_a, batches);
+	++tasks_remaining;
 
-	while (!stopped)
+	for (unsigned batch = 0; batch < batches; ++batch)
 	{
-		if (!checks.empty ())
-		{
-			auto check (checks.front ());
-			checks.pop_front ();
-			lock.unlock ();
-			verify (check);
-			condition.notify_all ();
-			lock.lock ();
-		}
-		else
+		int size = batch_size;
+		int start_index = batch * batch_size;
+
+		// Clamp it to the max number of verifications
+		if (start_index + batch_size > check_a.size)
 		{
-			condition.wait (lock);
+			size = overflow;
 		}
+
+		boost::asio::post (thread_pool, [this, task, size, start_index, batch] {
+			auto result = verify_batch (task->check, start_index, size);
+			release_assert (result);
+
+			if (--task->pending == 0)
+			{
+				task->check.promise->set_value ();
+				--tasks_remaining;
+			}
+		});
+	}
+}
+
+// Set the names of all the threads in the thread pool for easier identification
+void nano::signature_checker::set_name_threads (unsigned num_threads)
+{
+	bool ready = false;
+	int pending = num_threads;
+	std::condition_variable cv;
+	std::mutex mutex_l;
+	std::unique_lock<std::mutex> lk (mutex_l);
+
+	for (int i = 0; i < num_threads; ++i)
+	{
+		boost::asio::post (thread_pool, [&cv, &ready, &pending, &mutex_l]() {
+			nano::thread_role::set (nano::thread_role::name::signature_checking);
```

Comment:
This doesn't appear to work properly, only 2 out of 4 threads are named in my test run - probably because ::post picks the same two threads in the loop.

Reply:
Would be nice to make it work, thread tracking was a v17 feature and it may help diagnose issues in production.

Reply:
Ah ok, thanks for informing me 👍. Can you try the latest code and see if it works on yours now? (and more importantly actually exits the function); I tried it on windows/mac and it works ok even with 128 threads. Each thread now waits on a condition variable when reached, and the last one wakes them up which sets a promise so that we now when all threads have actually finished. 

I couldn't find anything that says calling `get_future()` on a `promise` is thread safe so I now get it before any threads are called.

I changed some inconsistent size_t/unsigned types which were causing build warnings

Reply:
Works fine here

Reply:
@cryptocode Thanks for checking